### ISSUE TITLE
AI client(?) creates default_config.ini in a potentially non writable location.

### ISSUE THREAD

**Issue by adrianbroher** [freeorion/freeorion#462]
Some part of the application writes a file called `default_config.ini` in the `default/AI` subdirectory.

I would like to point out that the whole 'default' subdirectory is usually not writable for users of the game.  Think of FreeOrion installations in the `/usr/`, `/application/`, `C:\Program Files` or similar directories.  Those directories are only writable for super users (root, Administrator) during installation time of the game.  When you need to write configuration data please use a user writable directory.

 The c++ `GetUserDir`function located in `util/Directories.cpp` gives you a path, that is intended for user specific configuration and save data, so please use this function and its Python binding (if available).

**Comment by Cjkjvfnby** [comment]
There is not python binding for `GetUserDir`
If someone add binding for it I will make python part.

Binding should looks like this

``` c++
def("getUserDir", Directories.GetUserDir, return_value_policy<return_by_value>(), "Return path to user directory, AI have write permission to this directory.");
```

**Comment by Vezzra** [comment]
Will try to get to that tomorrow.

**Comment by Vezzra** [comment]
@Cjkjvfnby, implementation of Python binding for GetUserDir committed: 162e8cd92729789da6f0881952276e852e292bde. Had to do it a bit differently than you suggested (for whatever reason returning the path to Python as std::string crashes the AI client, returning as boost::python:str however works).

You can go ahead and implement the required adjustments on the Python side. Re-assigned this issue to you.

**Comment by Vezzra** [comment]
On a side note: when storing the AI config file (`default_config.ini`) in this folder for user specific data, please rename it to something that indicates it's an **AI** config file. That will make its purpose clearer, as that location isn't for AI data only.

**Comment by Cjkjvfnby** [comment]
@Vezzra can you merge `python/AI/AIWrapper.cpp` changes to master?

**Comment by Vezzra** [comment]
@Cjkjvfnby: um, I already have? Commit 162e8cd92729789da6f0881952276e852e292bde has already been pushed to master... I can't see anything amiss?

**Comment by Cjkjvfnby** [comment]
My bad.

**Comment by Cjkjvfnby** [comment]
I try to ensure that only one AI write config file and this required a lot of changes.

We initialize python in 2 stages:
- import (and execute some code including config creation) 
- prepare AI in c++ code and call `initFreeOrionAI`

Can we prepare AI first and only after that import python libs?  This will make `initFreeOrionAI`  obsolete and make things less complicated.

**Comment by Cjkjvfnby** [comment]
Timers are dependent on config variables. I want config been used first time only on AI ready.
Here is example of code that allow to create timers on import (like it done now) and really initialize them only them AI is started.   

PS I write this implementation just to ensure that I did not forget how metaclasses works. 

``` python
from inspect import isfunction


class _Proxy(type):
    def __new__(cls, name, bases, arguments):
        arguments['_initialized'] = False

        init_function = arguments.pop('__init__')

        def init_on_call(method):
            def wrapper(self, *args, **kwargs):
                if not self._initialized:
                    self._init()
                    self._initialized = True
                return method(self, *args, **kwargs)
            return wrapper

        for k, v in arguments.items():
            if isfunction(v) and not k.startswith('__'):
                arguments[k] = init_on_call(v)

        def init(self, *args, **kwargs):
            init_function(self, *self.__args, **self.__kwargs)

        arguments['_init'] = init

        def store_init_arguments(self, *args, **kwargs):
            self.__args = args
            self.__kwargs = kwargs

        arguments['__init__'] = store_init_arguments
        return super(_Proxy, cls).__new__(cls, name, bases, arguments)


class Timer(object):
    __metaclass__ = _Proxy

    def __init__(self, arg):
        print " !! init called"
        self.arg = arg

    def start(self):
        return self.arg


print ">>> initialize"
f = Timer('hello')
print ">>> call"
print f.start()
print f.start()
```

```
>>> initialize
>>> call
 !! init called
hello
hello
```

**Comment by dbenage-cx** [comment]
Since I do not know the implications of this file, and there seems to be further issues, I stopped here and did not make a PR (`or "."` should probably be removed as well).

I realize there are more pressing concerns, but here are my questions/concerns once this is brought up later:
- Specifying `--ai-config fn`, the file is still created. util/MutliplayerCommon.cpp:41 seems to revert this to always false, so even with that option the log states `ai-config not set`.
- Changing the timers and timers_dump to True, results in a log entry
  `[debug] AI : Chat Message to human: AI_Error: AI script error in "generateOrders": "[Errno 2] No such file or directory: 'timers/timer-01.txt'"`. So the correct directories for timers need to be appropriated as well.
- If this file is to be tied to a specific AI type (e.g. default), this should probably go into a directory such as `getUserDir() / "AI" / ai_type`.
- From the comments on util/MultiPlayerCommon.cpp:38 and the file this is created in residing in a freeorion_debug folder, this seems to only be for testing.  Should it still be created in `Release` builds?  Would a new `getTmpDir()` be better suited?

**Comment by Cjkjvfnby** [comment]
`--ai-config` was designed as feature that allows user to customize AI without knowledge of python and changing source code. 

If there is no config present we try to create dummy config, that will help user to understand what values can be changed. This is not work, because on many systems game is installed to write protected directory. 

Dot should be totally removed. `getUserDir` will always return path, this is path where logs stored.

Other thing is creating config file N times, whee N is number of AI in that game. I don't sure if AI executed in different thread now, but it might be good idea and I don't know how open file for write will behave on all supported platforms.

**Comment by geoffthemedio** [comment]
Each AI player is controlled by a separate process, each of which can have numerous threads running.

**Comment by adrianbroher** [comment]
Pr #598 fixes the issue for me.  Thanks for providing the fix @dbenage-cx.

**Comment by Vezzra** [comment]
@Cjkjvfnby, I still have this issue:

> I try to ensure that only one AI write config file and this required a lot of changes.
> 
> We initialize python in 2 stages:
> - import (and execute some code including config creation) 
> - prepare AI in c++ code and call `initFreeOrionAI`
> 
> Can we prepare AI first and only after that import python libs?  This will make `initFreeOrionAI`  obsolete and make things less complicated.

...you addressed in this discussion here ([see above](https://github.com/freeorion/freeorion/issues/462#issuecomment-183865197)) on my todo list. I assume that this problem should still be addressed, even with the issue here now closed? If so, can you create a separate issue for that?

**Comment by Cjkjvfnby** [comment]
Hi, we throw the baby out with the bathwater. config without `getAIConfigStr` have no sense.  Will fix it myself.
